File: python-prototype/simulation/intersection.py
from typing import Type

from simulation.constants import CAR_PASS_CLOCK_TICKS, ONE_ROUND_CLOCK_TICKS
from simulation.crossing_connection.base.base_crossing_connection import (
    BaseCrossingConnection,
)

from .config import Config
from .crossing_state import CrossingState
from .errors.simulation_end import SimulationEnd
from .switching_strategy.base.base_switching_strategy import BaseSwitchingStrategy


class Intersection:
    def __init__(
        self,
        CrossingConnectionImplementation: Type[BaseCrossingConnection],
        SwitchingStrategyImplementation: Type[BaseSwitchingStrategy],
    ):
        self.config = Config()
        self.crossing_state = CrossingState(
            CrossingConnectionImplementation=CrossingConnectionImplementation
        )
        self.switching_strategy: BaseSwitchingStrategy = (
            SwitchingStrategyImplementation()
        )

        self.sim_round_ct = 0

    def run(self):
        if self.crossing_state.check_if_there_are_any_cars():
            print(self.crossing_state)
            print('No more cars at the intersection 🚗')

            raise SimulationEnd()

        print(
            f'==================== Round {self.sim_round_ct}, grace counter: {self.config.round_ct} ===================='
        )
        print(self.crossing_state)

        # update lights
        winning_direction = self.switching_strategy.calculate_winning_direction(
            crossing_state=self.crossing_state
        )

        if self.config.round_ct >= ONE_ROUND_CLOCK_TICKS:
            self.crossing_state.change_lights(
                winning_direction=winning_direction,
            )

            self.config.round_ct = 0

        if self.config.round_ct >= CAR_PASS_CLOCK_TICKS:
            self.crossing_state.move_cars(
                cars_counter_dict=self.crossing_state.cars_counter_dict
            )

        self.config.round_ct += 1
        self.sim_round_ct += 1

File: intersection-api/src/crossingLogic/intersection.ts
import { CAR_PASS_ROUND_TICKS, LIGHTS_CHANGE_ROUND_TICKS } from './constants';

import { BaseCrossingConnection } from './crossingConnection/base';
import { BaseSwitchingStrategy } from './switchingStrategy/base';
import Car from './Car';
import Config from './config';
import CrossingState from './crossingState';
import Direction from './direction';
import SimulationEnd from './errors/simulationEnd';

export default class Intersection {
	#config: Config;
	#crossingState: CrossingState;
	#switchingStrategy: BaseSwitchingStrategy;

	#simRoundCt: number;

	constructor(
		CrossingConnectionImplementation: new () => BaseCrossingConnection,
		SwitchingStrategyImplementation: new () => BaseSwitchingStrategy
	) {
		this.#config = Config.instance;
		this.#crossingState = new CrossingState(
			CrossingConnectionImplementation
		);
		this.#switchingStrategy = new SwitchingStrategyImplementation();
		this.#simRoundCt = 0;
	}

	addCar(car: Car) {
		this.#crossingState.addCar(car);
	}

	addRandomCar() {
		this.#crossingState.addRandomCar();
	}

	/**
	 * Runs the next round
	 * @return `Car`s which left the intersection this round
	 */
	run(): Set<Car> {
		let carsExitedThisRound = new Set<Car>();

		if (this.#crossingState.checkIfThereAreAnyCars()) {
			console.log(this.#crossingState.toString());
			console.log('No more cars at the intersection 🚗');
			throw new SimulationEnd('Simulation has ended');
		}

		console.log(
			`==================== Round ${this.#simRoundCt}, grace counter: ${this.#config.roundCt} ====================`
		);
		console.log(this.#crossingState.toString());

		// update lights
		let winningDirection: Direction =
			this.#switchingStrategy.calculateWinningDirection(
				this.#crossingState
			);

		if (this.#config.roundCt >= LIGHTS_CHANGE_ROUND_TICKS) {
			this.#crossingState.changeLights(
				(winningDirection = winningDirection)
			);

			this.#config.roundCt = 0;
		}

		if (this.#config.roundCt >= CAR_PASS_ROUND_TICKS) {
			carsExitedThisRound = this.#crossingState.moveCars(
				this.#crossingState.carsStore
			);
		}

		this.#config.roundCt++;
		this.#simRoundCt++;

		return carsExitedThisRound;
	}

	printConfig() {
		console.log(' >> Intersection configuration printed << ');

		console.log(
			`Lights change (🔴 -> 🟡, 🟡 -> 🟢, 🟡 -> 🔴) lasts ${LIGHTS_CHANGE_ROUND_TICKS} round(s)`
		);
		console.log(
			`A car driver waits ${CAR_PASS_ROUND_TICKS} round(s) after 🟢 turns on before driving`
		);
		console.log(
			`The chosen BaseCrossingConnection implementation is: ${this.#crossingState.crossingConnection.constructor.name}, which does the following: ${this.#crossingState.crossingConnection.getDescription()}`
		);

		console.log(' >> ---------------------------------- << ');
	}
}
